File: UVA/Random/10651 - Pebble Solitaire.cpp
#include <assert.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <time.h>
#include <algorithm>
#include <numeric>
#include <functional>
#include <bitset>
#include <vector>
#include <list>
#include <set>
#include <map>
#include <queue>
#include <stack>
#include <string>
#include <sstream>
#include <iostream>
#include <limits.h>
using namespace std;

typedef long long int64;
#define MAX 12

int dp[1 << 13];

int solve(int mask) {
	if (mask <= 0) return 0;
	if (dp[mask] != -1) return dp[mask];
	dp[mask] = INT_MAX;
	bool cant = true;
	for (int a = 0; a < MAX; ++a) {
		if (a + 2 < MAX) {
			if ((mask & (1 << a)) && (mask & (1 << (a + 1))) && !(mask & (1 << (a + 2)))) {
				cant = false;
				int mask1 = mask;
				mask1 |= (1 << (a + 2));
				mask1 &= ~(1 << (a + 1));
				mask1 &= ~(1 << a);
				dp[mask] = min(dp[mask], solve(mask1));
			}
		}
		if (a - 1 >= 0) {
			if ((mask & (1 << a)) && (mask & (1 << (a + 1))) && !(mask & (1 << (a - 1)))) {
				cant = false;
				int mask1 = mask;
				mask1 |= (1 << (a - 1));
				mask1 &= ~(1 << (a + 1));
				mask1 &= ~(1 << a);
				dp[mask] = min(dp[mask], solve(mask1));
			}
		}
	}
	int cnt = 0;
	if (cant) {
		for (int a = 0; a < MAX; ++a)
			if (mask & (1 << a)) ++cnt;
		return dp[mask] = min(dp[mask], cnt);
	}
	return dp[mask];
}

int main() {
#ifndef ONLINE_JUDGE
	freopen("in.txt", "r", stdin);
#endif
	int cases; cin >> cases;
	while (cases--) {
		string str; cin >> str;
		int mask = 0;
		for (int a = 0; a < str.length(); ++a)
			if (str[a] == 'o') mask |= (1 << a);
		memset(dp, -1, sizeof(dp));
		cout << solve(mask) << endl;
	}
	return 0;
}

File: UVA/Random/10651 - Pebble Solitaire.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Arrays;
import java.util.Scanner;

public class Main {
    private static void debug(Object...os) {
        System.out.println(Arrays.deepToString(os));
    }

    static int[] cache = new int[1 << 12];

    public static void main(String[] args) {
        Scanner cin = new Scanner(System.in);
        int T = cin.nextInt();
        while (T-- != 0) {
            String str = cin.next();
            int mask = 0;
            for (int i = 0; i < str.length(); ++i)
                if (str.charAt(i) == 'o')
                    mask ^= (1 << i);
            Arrays.fill(cache, -1);
            System.out.println(solve(mask));
        }
    }

    private static int solve(int mask) {
        if (mask <= 0) return 0;
        if (cache[mask] != -1)
            return cache[mask];
        cache[mask] = Integer.MAX_VALUE;
        boolean cant = true;
        for (int i = 0; i < 12; ++i) {
            if (i + 2 < 12 && ok(i, i + 1, i + 2, mask)) {
                cache[mask] = Math.min(cache[mask], solve(mask ^ (1 << i) ^ (1 << (i + 1)) ^ (1 << (i + 2))));
                cant = false;
            }
            if (i - 1 >= 0 && ok(i, i + 1, i - 1, mask)) {
                cache[mask] = Math.min(cache[mask], solve(mask ^ (1 << i) ^ (1 << (i + 1)) ^ (1 << (i - 1))));
                cant = false;
            }
        }
        if (cant) {
            int cnt = 0;
            for (int i = 0; i < 12; ++i)
                if ((mask & (1 << i)) != 0)
                    ++cnt;
            cache[mask] = Math.min(cache[mask], cnt);
        }
        return cache[mask];
    }

    private static boolean ok(int a, int b, int c, int mask) {
        return (mask & (1 << a)) != 0 && (mask & (1 << b)) != 0 && (mask & (1 << c)) == 0;
    }
}